### ISSUE TITLE
campaigns only sending 1 at a time

### ISSUE THREAD

**Issue by cyb456** [bramley/phplist-plugin-autoresponder#10]
After updating to phplist 3.6.7 autoresponders campaigns only sends 1 email at a time every time phplist -p processqueue is run, there are hundreds waiting so it should do them all at once

**Comment by bramley** [comment]
From which phplist version did you upgrade? Which php version are you using?

If you view the autoresponder, what does that show for the number of subscribers ready?

**Comment by cyb456** [comment]
from 3.6.6, using PHP 7.0.33-0ubuntu0.16.04.16

It says 860 or so ready

```phplist -p processqueue
phpList - phpList version 3.6.7 (c) 2000-2022 phpList Ltd, https://www.phplist.com
phpList - Recently sent : 0
phpList - Started [0.0352810000] (213)
phpList - Sending in batches of 20 emails [0.0002870000] (214)
phpList - Processing has started, [0.0005270000] (217)
phpList - One campaign to process. [0.0011060000] (219)
phpList - Processing campaign 8 [0.0158220000] (257)
phpList - Looking for subscribers [0.0018200000] (262)
phpList - Found them: 1 to process [0.0051880000] (268)
phpList - Processed 1 out of 1 subscribers [3.0259550000] (316)
phpList - It took  96 days 17 hours 5 mins 26 seconds to send this message [0.0044770000] (336)
phpList - Script stage: 5 [0.0009700000] (338)
phpList - 1 messages sent in 3.09 seconds (1164 msgs/hr) [0.0004610000] (341)
phpList - Finished this run [0.0013760000] (347)
```
I tried after this turning off automatic processing, and tried sending the campaigns manually from the web panel, unfortunately it then sent ALL of the campaigns at once instead of the week interval between them for subscribers that were past due

**Comment by bramley** [comment]
> `phpList - Found them: 1 to process [0.0051880000] (268)`

This seems to indicate that phplist found only one subscriber to be sent the campaign, regardless of this plugin. i.e. only one subscriber who is on the list, confirmed and not blacklisted, and not already sent the campaign. I would expect that the figure should show the 860 you mentioned.

> I tried after this turning off automatic processing, and tried sending the campaigns manually from the web panel, unfortunately it then sent ALL of the campaigns at once instead of the week interval between them for subscribers that were past due

I don't understand this. Processing the queue through the browser should work the same way as the command line.

**Comment by bramley** [comment]
Can you give me access to look at your phplist? You can send a message with the access details at https://www.dcameron.me.uk/contact/

**Comment by cyb456** [comment]
I cannot do that sorry

**Comment by bramley** [comment]
To try to get some more information on what is happening can you try these

enable VERBOSE in config.php

    define('VERBOSE', 1);

There will be more detail in the output of processqueue.

Enable logging for the plugin, add this to config.php

```
$log_options = [
    'threshold' => 'DEBUG',
    'classes' => [
        'Autoresponder' => true,
    ]
];
```
Log files will be created in the phplist $tmpdir directory.
The Autoresponder plugin should log each autoresponder and how many subscribers are ready to be sent the campaign.

Also, are you using any other plugin that selects subscribers to be sent to?

**Comment by cyb456** [comment]
I did phplist -pprocessqueue twice today, once at 10:09 and once at 10:25
campaign id 7 is the first, then they get passed on to campaign id 8 for the next email in a week

As you can see each time it is only sending 1 email to the second campaign, (and probably the third but no emails waiting there at the moment). I had to run the "phplist -pprocessqueue" command 5 more times in a row to empty out the subscribers waiting
```
[Wed 20 Apr 2022 10:09:01] [debug] Autoresponder::processQueueStart, line 98
7 119
[Wed 20 Apr 2022 10:09:01] [debug] Autoresponder::processQueueStart, line 103
Campaign 7 submitted
[Wed 20 Apr 2022 10:09:01] [debug] Autoresponder::processQueueStart, line 98
8 7
[Wed 20 Apr 2022 10:09:01] [debug] Autoresponder::processQueueStart, line 103
Campaign 8 submitted
[Wed 20 Apr 2022 10:09:01] [debug] Autoresponder::processQueueStart, line 98
9 0

[Wed 20 Apr 2022 10:25:26] [debug] Autoresponder::processQueueStart, line 98
7 0
[Wed 20 Apr 2022 10:25:26] [debug] Autoresponder::processQueueStart, line 98
8 6
[Wed 20 Apr 2022 10:25:26] [debug] Autoresponder::processQueueStart, line 103
Campaign 8 submitted
[Wed 20 Apr 2022 10:25:26] [debug] Autoresponder::processQueueStart, line 98
9 0
```
10:09AM
```
phpList - phpList version 3.6.7 (c) 2000-2022 phpList Ltd, https://www.phplist.com
phpList - Maximum time for queue processing: 99999 [0.0720540000] (253)
phpList - Recently sent : 0
phpList - Started [0.0042810000] (256)
phpList - Sending in batches of 1 emails [0.0002730000] (257)
phpList -  select id from phplist_message where status not in ("draft", "sent", "prepared", "suspended") and embargo  [0.0002260000] (258)
phpList - Processing has started, [0.0012600000] (261)
phpList - 2 campaigns to process. [0.0006700000] (263)
phpList - sending of this campaign will stop, if it is still going in  10 years 258 days 13 hours 50 mins 59 seconds [0.0201950000] (318)
phpList - Processing campaign 7 [0.0041370000] (326)
phpList - Looking for subscribers [0.0027330000] (331)
phpList - User select query select distinct u.id from phplist_listuser as listuser
        inner join phplist_user_user as u ON u.id = listuser.userid
        inner join phplist_listmessage as listmessage ON listuser.listid = listmessage.listid
        left join phplist_usermessage as um ON (um.messageid = 7 and um.userid = listuser.userid)
        where
        listmessage.messageid = 7
        and listmessage.listid = listuser.listid
        and u.id = listuser.userid
        and um.userid IS NULL
        and u.confirmed and !u.blacklisted and !u.disabled
          [0.0007780000] (334)
phpList - Found them: 119 to process [0.0064890000] (338)
phpList - Sending 7 to email@host [0.0072140000] (349)
phpList - It took 0.0196080000 seconds to send [0.0205670000] (385)
^repeats for each till 119 (this is working correctly these were all newly added in past 24 hours)

phpList - Processed 119 out of 119 subscribers [0.0025060000] (4051)
phpList - It took  23 hours 50 mins 04 seconds to send this message [0.0028740000] (4060)
phpList - sending of this campaign will stop, if it is still going in  10 years 258 days 13 hours 50 mins 56 seconds [0.0122050000] (4093)
phpList - Processing campaign 8 [0.0014980000] (4098)
phpList - Looking for subscribers [0.0014620000] (4102)
phpList - User select query select distinct u.id from phplist_listuser as listuser
        inner join phplist_user_user as u ON u.id = listuser.userid
        inner join phplist_listmessage as listmessage ON listuser.listid = listmessage.listid
        left join phplist_usermessage as um ON (um.messageid = 8 and um.userid = listuser.userid)
        where
        listmessage.messageid = 8
        and listmessage.listid = listuser.listid
        and u.id = listuser.userid
        and um.userid IS NULL
        and u.confirmed and !u.blacklisted and !u.disabled
          [0.0005500000] (4105)
phpList - Found them: 120 to process [0.0052470000] (4107)
phpList - Sending 8 to sales@comapny [0.0024740000] (4115)
phpList - It took 0.0120710000 seconds to send [0.0123920000] (4133)
phpList - sent 120 ETA Wed 20 Apr 10:09 sending 152952 msg/hr
^valid send, first campaign was sent a week ago
the next "6 waiting" should be right here!

phpList - Sending blocked by plugin Autoresponder
phpList - Not sending email@host [0.0053470000] (4147)
phpList - sent 120 ETA Wed 20 Apr 10:09 sending 152716 msg/hr
^ repeats 119 times, these were the 119 just sent the first campaign

phpList - Processed 120 out of 120 subscribers [0.0024050000] (5688)
phpList - It took  23 hours 50 mins 02 seconds to send this message [0.0043180000] (5695)
phpList - Script stage: 5 [0.0017350000] (5697)
phpList - 120 messages sent in 3.33 seconds (129919 msgs/hr) [0.0004440000] (5700)
phpList - Finished this run [0.0011800000] (5706)
```
10:25AM
```
phpList - phpList version 3.6.7 (c) 2000-2022 phpList Ltd, https://www.phplist.com
phpList - Maximum time for queue processing: 99999 [0.0382670000] (220)
phpList - Recently sent : 0
phpList - Started [0.0021480000] (223)
phpList - Sending in batches of 1 emails [0.0002000000] (224)
phpList -  select id from phplist_message where status not in ("draft", "sent", "prepared", "suspended") and embargo  [0.0006170000] (225)
phpList - Processing has started, [0.0011490000] (228)
phpList - One campaign to process. [0.0005240000] (230)
phpList - sending of this campaign will stop, if it is still going in  10 years 258 days 13 hours 37 mins 30 seconds [0.0164500000] (285)
phpList - Processing campaign 8 [0.0028730000] (293)
phpList - Looking for subscribers [0.0019900000] (298)
phpList - User select query select distinct u.id from phplist_listuser as listuser
        inner join phplist_user_user as u ON u.id = listuser.userid
        inner join phplist_listmessage as listmessage ON listuser.listid = listmessage.listid
        left join phplist_usermessage as um ON (um.messageid = 8 and um.userid = listuser.userid)
        where
        listmessage.messageid = 8
        and listmessage.listid = listuser.listid
        and u.id = listuser.userid
        and um.userid IS NULL
        and u.confirmed and !u.blacklisted and !u.disabled
          [0.0009300000] (301)
phpList - Found them: 1 to process [0.0058080000] (305)
phpList - Sending 8 to email@host [0.0059520000] (316)
phpList - It took 0.0229910000 seconds to send [0.0238580000] (352)
phpList - Processed 1 out of 1 subscribers [0.0040030000] (360)
phpList - It took  24 hours 3 mins 27 seconds to send this message [0.0041200000] (369)
phpList - Script stage: 5 [0.0005340000] (371)
phpList - 1 messages sent in 0.11 seconds (32888 msgs/hr) [0.0004900000] (374)
phpList - Finished this run [0.0016470000] (380)
```

plugins:
```
Autoresponder 3.3.2+20220218
CampaignsPlugin 2.4.0+20200216
CKEditorPlugin 2.6.2+20210428
CommonPlugin 3.18.0+20211228
dateplaceholder 0.2
listcleaner 0.2
MessageStatisticsPlugin 2.3.6+20211231
SegmentPlugin 2.12.1+20210428
simpleattributeselect 0.1
subjectLinePlaceholdersPlugin 1.0a4
SubscribersPlugin 2.31.0+20220112
ViewBrowserPlugin 2.11.5+20220204
```

**Comment by bramley** [comment]
Thanks.
For campaign 8 at 10:09 the plugin found 7 subscribers ready to be sent.
phplist found the 120 subscribers who are on the list, but only one of the 7 was actually sent, the other 6 and the 113 who were not ready were blocked by the plugin.

At 10:25 the plugin found there were 6 subscriber ready, but phplist found only 1.

What are the autoresponder periods for the two campaigns 7 and 8?
Are you using the Segment plugin with these campaigns?

**Comment by cyb456** [comment]
5 minutes after added to list for campaign id7 (which then throws emails to another list)
6 days after added to that next list for campaign id 8

I am not using the segment plugin

**Comment by bramley** [comment]
Can you update the plugin, I have added logging of the subscribers selected by the plugin.
Use this URL on the Manage Plugins page
 `https://github.com/bramley/phplist-plugin-autoresponder/archive/debug_ready_subscribers.zip`

If you can run processqueue again when there are some subscribers ready. If not all of them are sent, which seems to be likely, the please try to identify some that should have been sent but weren't, i.e. belong to the right list, time added to the list, confirmed, etc. I'd like to see whether they are in the group selected by the plugin.

Also, can you revert to using a bigger batch size. I'm not sure that a batch size of 1 helps to clarify what is happening, and might introduce other problems.

**Comment by bramley** [comment]
I found the cause of the problem after trying to exactly mimic your process of two autoresponders. You need to upgrade both this plugin and Common Plugin on the Manage Plugins page. Ignore the "debug_ready_subscribers"  branch mentioned above.

**Comment by cyb456** [comment]
Strange, I had set the batch size from 20 to commented out so it should have been unlimited, maybe just displays 1?

I don't have any ready to get 2nd email today, but I will be able to confirm the fix monday morning

**Comment by cyb456** [comment]
I believe it is fixed, thank you!

**Comment by bramley** [comment]
Thanks, you can remove the logging and verbose settings if you have not already done so.
